File: analysis/execucao_orcamentaria.py
from typing import Any

import pandas as pd
from sqlalchemy import text


def run(conn: Any, year: int, empresa_ids: list[str] | None = None) -> pd.DataFrame:
    empresa_clause = "AND empresa = ANY(:empresas)" if empresa_ids else ""
    params: dict = {"ano": year}
    if empresa_ids:
        params["empresas"] = empresa_ids
    df = pd.read_sql_query(
        text(
            f"SELECT ano, empresa, codigo, descricao, empenhado, liquidado, pago, dotacao_atualizada FROM despesas_por_orgao WHERE ano = :ano {empresa_clause}"
        ),
        conn,
        params=params,
    )
    df["empenhado"] = pd.to_numeric(df["empenhado"].str.replace(",", "."), errors="coerce").fillna(0)
    df["liquidado"] = pd.to_numeric(df["liquidado"].str.replace(",", "."), errors="coerce").fillna(0)
    df["pago"] = pd.to_numeric(df["pago"].str.replace(",", "."), errors="coerce").fillna(0)
    df["dotacao_atualizada"] = pd.to_numeric(df["dotacao_atualizada"].str.replace(",", "."), errors="coerce").fillna(0)
    df["taxa_execucao"] = df.apply(
        lambda r: r["empenhado"] / r["dotacao_atualizada"] if r["dotacao_atualizada"] > 0 else 0,
        axis=1,
    )
    df["alerta"] = df.apply(
        lambda r: (
            "N/D"
            if r["empenhado"] == 0 and r["dotacao_atualizada"] == 0 and r["taxa_execucao"] == 0
            else ("baixa" if r["taxa_execucao"] < 0.3 else ("excesso" if r["taxa_execucao"] > 1.0 else "normal"))
        ),
        axis=1,
    )
    return df


def summarize(df: pd.DataFrame) -> dict:
    total_dotacao = float(df["dotacao_atualizada"].sum())
    total_empenhado = float(df["empenhado"].sum())
    total_liquidado = float(df["liquidado"].sum())
    total_pago = float(df["pago"].sum())
    return {
        "total_empenhado": total_empenhado,
        "total_liquidado": total_liquidado,
        "total_pago": total_pago,
        "total_dotacao": total_dotacao,
        "saldo_orcamentario": total_dotacao - total_empenhado,
    }


def top_orgaos_por_dotacao(df: pd.DataFrame, n: int = 10) -> pd.DataFrame:
    """Agregar por órgão, calcular taxa de execução + alerta, retornar top-n por dotação."""
    por_orgao = df.groupby(["empresa", "descricao"], as_index=False).agg(
        empenhado=("empenhado", "sum"), dotacao_atualizada=("dotacao_atualizada", "sum")
    )
    por_orgao["taxa_execucao"] = por_orgao.apply(
        lambda r: r["empenhado"] / r["dotacao_atualizada"] if r["dotacao_atualizada"] > 0 else 0.0,
        axis=1,
    )
    por_orgao["alerta"] = por_orgao.apply(
        lambda r: (
            "N/D"
            if r["empenhado"] == 0 and r["dotacao_atualizada"] == 0
            else ("baixa" if r["taxa_execucao"] < 0.3 else ("excesso" if r["taxa_execucao"] > 1.0 else "normal"))
        ),
        axis=1,
    )
    return por_orgao.nlargest(n, "dotacao_atualizada").copy()


def summarize_by_year(conn: Any, years: list[int], empresa_ids: list[str] | None = None) -> dict[int, dict]:
    return {year: summarize(run(conn, year, empresa_ids=empresa_ids)) for year in years}

File: packages/db/src/queries/execucao_orcamentaria.ts
import { sql } from "kysely";
import { db } from "../client";

export interface ItemExecucaoOrcamentaria {
  ano: number;
  empresa: string;
  codigo: string;
  descricao: string;
  empenhado: number;
  liquidado: number;
  pago: number;
  dotacao_atualizada: number;
  taxa_execucao: number;
  alerta: "N/D" | "baixa" | "excesso" | "normal";
}

export interface ExecucaoSummary {
  total_empenhado: number;
  total_liquidado: number;
  total_pago: number;
  total_dotacao: number;
  saldo_orcamentario: number;
}

export async function getExecucaoOrcamentaria(
  year: number,
  empresaIds?: string[] | null,
): Promise<ItemExecucaoOrcamentaria[]> {
  try {
    let q = sql`
      SELECT ano, empresa, codigo, descricao, empenhado, liquidado, pago, dotacao_atualizada
      FROM fct_despesas_por_orgao
      WHERE ano = ${year}
    `;
    if (empresaIds && empresaIds.length > 0) {
      q = sql`${q} AND empresa = ANY(${empresaIds})`;
    }
    const res = await q.execute(db);
    const rows = (res.rows as any[]) || [];

    return rows.map((r) => {
      const emp = parseFloat(String(r.empenhado ?? "0").replace(",", ".")) || 0;
      const liq = parseFloat(String(r.liquidado ?? "0").replace(",", ".")) || 0;
      const pag = parseFloat(String(r.pago ?? "0").replace(",", ".")) || 0;
      const dot =
        parseFloat(String(r.dotacao_atualizada ?? "0").replace(",", ".")) || 0;
      const taxa = dot > 0 ? emp / dot : 0;

      let alerta: "N/D" | "baixa" | "excesso" | "normal" = "normal";
      if (emp === 0 && dot === 0 && taxa === 0) {
        alerta = "N/D";
      } else if (taxa < 0.3) {
        alerta = "baixa";
      } else if (taxa > 1.0) {
        alerta = "excesso";
      }

      return {
        ano: Number(r.ano),
        empresa: String(r.empresa ?? ""),
        codigo: String(r.codigo ?? ""),
        descricao: String(r.descricao ?? ""),
        empenhado: emp,
        liquidado: liq,
        pago: pag,
        dotacao_atualizada: dot,
        taxa_execucao: taxa,
        alerta,
      };
    });
  } catch {
    return [];
  }
}

export function summarizeExecucao(
  items: ItemExecucaoOrcamentaria[],
): ExecucaoSummary {
  const total_empenhado = items.reduce((acc, i) => acc + i.empenhado, 0);
  const total_liquidado = items.reduce((acc, i) => acc + i.liquidado, 0);
  const total_pago = items.reduce((acc, i) => acc + i.pago, 0);
  const total_dotacao = items.reduce((acc, i) => acc + i.dotacao_atualizada, 0);
  return {
    total_empenhado,
    total_liquidado,
    total_pago,
    total_dotacao,
    saldo_orcamentario: total_dotacao - total_empenhado,
  };
}
